### PR TITLE
Investigate noIterationTypes crash

### LINKED ISSUE #32953: Language service frequently stops in these days
Language service warns wrong errors based on old code and I have to restart language service.

**TypeScript Version:**  3.4.0-dev.20190817

VSCode 1.37.1

```
[2019-08-18 01:55:51.644] [exthost] [error] Error: <semantic> TypeScript Server Error (3.7.0-dev.20190817)
Not supported
Error: Not supported
    at Object.get yieldType [as yieldType] (tsserver.js:32546:37)
    at getIteratedTypeOrElementType (tsserver.js:57722:61)
    at getContextualTypeForElementExpression (tsserver.js:49558:20)
    at checkArrayLiteral (tsserver.js:50019:49)
    at checkExpressionWorker (tsserver.js:55222:28)
    at checkExpression (tsserver.js:55147:38)
    at checkFunctionExpressionOrObjectLiteralMethodDeferred (tsserver.js:53886:36)
    at checkDeferredNode (tsserver.js:59994:21)
    at Map.forEach (<anonymous>)
    at checkDeferredNodes (tsserver.js:59982:37)
    at checkSourceFileWorker (tsserver.js:60044:17)
    at checkSourceFile (tsserver.js:60014:13)
    at getDiagnosticsWorker (tsserver.js:60089:17)
    at Object.getDiagnostics (tsserver.js:60075:24)
    at isMissingAwaitError (tsserver.js:116089:39)
    at getAwaitableExpression (tsserver.js:116110:20)
    at Object.getCodeActions (tsserver.js:116050:34)
    at tsserver.js:115926:121
    at Object.flatMap (tsserver.js:583:25)
    at Object.getFixes (tsserver.js:115926:23)
    at tsserver.js:125977:35
    at Object.flatMap (tsserver.js:583:25)
    at Object.getCodeFixesAtPosition (tsserver.js:125975:23)
    at IOSession.Session.getCodeFixes (tsserver.js:135174:64)
    at Session.handlers.ts.createMapFromTemplate._a.(anonymous function) (tsserver.js:133961:61)
    at tsserver.js:135367:88
    at IOSession.Session.executeWithRequestId (tsserver.js:135358:28)
    at IOSession.Session.executeCommand (tsserver.js:135367:33)
    at IOSession.Session.onMessage (tsserver.js:135390:35)
    at Interface.<anonymous> (tsserver.js:136705:27)
    at Interface.emit (events.js:182:13)
    at Interface._onLine (readline.js:290:10)
    at Interface._normalWrite (readline.js:433:12)
    at Socket.ondata (readline.js:149:10)
    at Socket.emit (events.js:182:13)
    at addChunk (_stream_readable.js:283:12)
    at readableAddChunk (_stream_readable.js:264:11)
    at Socket.Readable.push (_stream_readable.js:219:10)
    at Pipe.onStreamRead [as onread] (internal/stream_base_commons.js:94:17)
	at Function.create (c:\Users\\AppData\Local\Programs\Microsoft VS Code\resources\app\extensions\typescript-language-features\dist\extension.js:53:63463)
	at h.dispatchResponse (c:\Users\\AppData\Local\Programs\Microsoft VS Code\resources\app\extensions\typescript-language-features\dist\extension.js:53:56618)
	at h.dispatchMessage (c:\Users\\AppData\Local\Programs\Microsoft VS Code\resources\app\extensions\typescript-language-features\dist\extension.js:53:55516)
	at constructor._reader.onData.e (c:\Users\\AppData\Local\Programs\Microsoft VS Code\resources\app\extensions\typescript-language-features\dist\extension.js:53:54992)
	at u.fire (c:\Users\\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:50:207)
	at t.Reader.onLengthData (c:\Users\\AppData\Local\Programs\Microsoft VS Code\resources\app\extensions\typescript-language-features\dist\extension.js:53:61683)
	at Socket.t.Reader.constructor.e.on.e (c:\Users\\AppData\Local\Programs\Microsoft VS Code\resources\app\extensions\typescript-language-features\dist\extension.js:53:61331)
	at Socket.emit (events.js:182:13)
	at addChunk (_stream_readable.js:283:12)
	at readableAddChunk (_stream_readable.js:264:11)
	at Socket.Readable.push (_stream_readable.js:219:10)
	at Pipe.onStreamRead [as onread] (internal/stream_base_commons.js:94:17)
```

```
[2019-08-17 16:27:42.203] [exthost] [error] Error: DisposableStore is missing...
	at R._convertCompletionItem (c:\Users\\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:563:173)
	at define.resolveCompletionItem.o.asPromise.then.e (c:\Users\\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:562:707)
```

### PR DESCRIPTION
**Pull request by rbuckton**
Investigating a crash related to the `noIterationTypes` sentinel.

Fixes #32953

### CODE REVIEW

**Comment by rbuckton** [comment]
@typescript-bot pack this

**Comment by rbuckton** [comment]
@typescript-bot pack this

**Comment by rbuckton** [comment]
@typescript-bot pack this

**Comment by rbuckton** [comment]
@typescript-bot pack this

**Comment by rbuckton** [comment]
@dbaeumer, @falsandtru: Once @typescript-bot indicates the tarball for the proposed fix has completed, would you be able to try it out in your local workspace and report back as to whether this has improved stability in the editor?

**Comment by falsandtru** [comment]
Looks like the problem disappeared :+1:

**Comment by rbuckton** [comment]
I'd like to verify whether this issue was indeed related to `nameType` being incorrectly set on a `Symbol`. After the bot finishes packing up the new changes, I'd appreciate if you were able to try the new tarball to see if the issue is still resolved.

@typescript-bot pack this

**Comment by falsandtru** [comment]
I can repro with the reverted version. As I wrote in #32953, I can repro also with yesterday's build.

**Comment by rbuckton** [comment]
Unfortunately, I still am unable to repro this issue following the steps you provided.

**Comment by rbuckton** [comment]
@typescript-bot pack this

**Comment by rbuckton** [comment]
@falsandtru, @dbaeumer: Unfortunately I've been unable to repro this using either @falsandtru's repro steps or by manually making changes to the ESLint VSCode extension that @dbaeumer was working on in their issue.

While it looks like we have [verified](https://github.com/microsoft/TypeScript/pull/36388#issuecomment-577965942) that we can resolve this specific issue by moving `noIterationTypes` outside of `createTypeChecker`, the fact that we could end up with multiple *different* versions of `noIterationTypes` indicates there is a leak somewhere in the type checker, as this should not be possible.

So far I have identified 3 possible culprits that could leak a `TypeChecker` (and an older `noIterationTypes`):

- The `nameType` property on `Symbol`, which could leak a type on symbols created by the binder that could be reused during incremental parse.
  - We have already verified that this was [not the underlying cause](https://github.com/microsoft/TypeScript/pull/36388#issuecomment-578026816) in @falsandtru's case.
- The `contextualType` property on `Node` could outlive a `TypeChecker` if the current semantic operation is canceled (via a cancellation token).
- The `combineValueAndTypeSymbols` function could possibly assign a `SymbolTable` created by the binder to a `TransientSymbol` created by the checker. This could also result in a leak as much of the checker assumes that symbol tables on transient symbols are themselves transient and won't outlive the checker.

I've just pushed up a commit that adds a number of checks to the `TypeChecker` public API and several internal APIs, as well as addressed the issues mentioned above. The hope is that either one of the above issues was the culprit, or that we will be able to use the additional checks to narrow down the underlying leak. If you are able, once the bot has finished packing a tarball of these changes, I would appreciate if you could use it to test your respective scenarios.

**Comment by falsandtru** [comment]
Not reproducible as expected.

**Comment by rbuckton** [comment]
The issue no longer exists as of this build? That indicates the issue was one of the three mentioned above.

**Comment by falsandtru** [comment]
Yes. The build for debugging should output debug log to discover the problem.

**Comment by falsandtru** [comment]
Another, the log of https://github.com/microsoft/TypeScript/issues/32953#issuecomment-577958476 doesn't help to investigate how to repro?

**Comment by dbaeumer** [comment]
@rbuckton we have endgame this week so not too much programming. But I will make sure I use the latest build and have an eye on stale problems. From my experience I think I got hit by the cancellation. The traces even showed that a semantic diagnostic operation got canceled before. So may it can be reproduce by setting a break point into the server that results in the client to cancel the operation.

**Comment by rbuckton** [comment]
Unfortunately I don't have a test that can repro this behavior, so I am hoping at the very least that the guards that I've added will be enough to catch future leaks of this kind.

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -317,21 +317,6 @@ namespace ts {
         const emitResolver = createResolver();
         const nodeBuilder = createNodeBuilder();
 
-        const globals = createSymbolTable();
```
Comment (rbuckton) [inline src/compiler/checker.ts]: NOTE: I needed to move these after the `const checker = {...}` declaration as `createSymbol` now depends on `checker` to ensure we add the current `TypeChecker` to all symbols to make sure we are always using the correct checker.

**Comment by rbuckton** [comment]
@typescript-bot perf test

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -351,10 +336,14 @@ namespace ts {
             isUndefinedSymbol: symbol => symbol === undefinedSymbol,
             isArgumentsSymbol: symbol => symbol === argumentsSymbol,
             isUnknownSymbol: symbol => symbol === unknownSymbol,
-            getMergedSymbol,
+            getMergedSymbol: symbol => {
+                verifyTypeCheckerOfSymbol(symbol);
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: How about having `verifyCheckerOfXXX` return the value it is passed such that you can write `return getMergedSymbol(verifyCheckerOfSymbol(symbol))` and keep it to just one expression.

File: src/compiler/utilities.ts [hunk 3]
```diff
@@ -4972,20 +4972,17 @@ namespace ts {
         this.id = undefined;
         this.mergeId = undefined;
         this.parent = undefined;
+        this.checker = undefined;
```
Comment (ahejlsberg) [inline src/compiler/utilities.ts]: Not crazy about this *always* being there.

**Comment by ahejlsberg** [comment]
I'm not totally sold on the value of the `verifyCheckerOfXXX` calls and the `checker` property always being present. It's overhead and the bug we're fixing wasn't actually caused by anything this is verifying.

**Comment by rbuckton** [comment]
I can pull the API verification for now, but we really should consider ensuring our public API doesn't allow you to try to mix and match types/signatures/symbols produces by other checkers.

**Comment by rbuckton** [comment]
@ahejlsberg: Any other comments after having reverted the `verifyX` changes for now?

File: src/compiler/checker.ts [hunk 4]
```diff
@@ -2277,8 +2278,8 @@ namespace ts {
             result.declarations = deduplicate(concatenate(valueSymbol.declarations, typeSymbol.declarations), equateValues);
             result.parent = valueSymbol.parent || typeSymbol.parent;
             if (valueSymbol.valueDeclaration) result.valueDeclaration = valueSymbol.valueDeclaration;
-            if (typeSymbol.members) result.members = typeSymbol.members;
-            if (valueSymbol.exports) result.exports = valueSymbol.exports;
+            if (typeSymbol.members) result.members = cloneMap(typeSymbol.members);
+            if (valueSymbol.exports) result.exports = cloneMap(valueSymbol.exports);
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: Why do we need to clone here?

Comment (weswigham) [reply src/compiler/checker.ts]: if `typeSymbol` or `valueSymbol` are _not_ `TransietSymbol`s, then their symbol tables consist only of non-transient symbols and persist through multiple checks. If we only copy the reference to the table into a new `TransientSymbol`, when we later stuff more things into the table (as is possible for transient symbols), we'll pollute the supposedly "immutable" symbol table from the original non-transient symbol with new transient members.

Comment (rbuckton) [reply src/compiler/checker.ts]: This was one of 3 possible leaks. We assume elsewhere in the checker that any `SymbolTable` attached to a `TransientSymbol` is safe to mutate, and sometimes add new symbols to it. `combineValueAndTypeSymbols` currently just blindly assigns the reference to the symbol table to `result` (which is a `TransientSymbol`), ignoring whether `typeSymbol` or `valueSymbol` were created by the binder. As a result, there is the possibility of some other function in checker later mutating the `SymbolTable`.

**Comment by rbuckton** [comment]
Need to reopen this on master since I used release-3.8 primarily for providing a stable drop for testers.

@typescript-bot cherry-pick this to master

**Comment by rbuckton** [comment]
Closing in favor of https://github.com/microsoft/TypeScript/pull/36491, which ports this change to `master` instead.

**Comment by rbuckton** [comment]
This has been merged into `master`.

**Pull request closed without merging**